randutil/rand.go

```diff
@@ -0,0 +1,95 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package randutil initialises properly random value generation and
+// exposes a streamlined set of functions for it, including for crypto
+// random tokens.
+package randutil
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"math/rand"
+	"net"
+	"os"
+	"sync"
+	"time"
+)
+
+func init() {
+	// golang does not init Seed() itself
+	rand.Seed(time.Now().UnixNano() + int64(os.Getpid()))
```

Comment:
@flotter @anpep Already discussed this with Fred online, but for the record, there's a reason why Go does not seed by default, and we shoud not be silently doing this inside a util library either, as it's taking away the chance of the application from controlling its own seed, in whatever way is appropriate for the given application. Often times, applications will also want to have a fixed seed for debugging purposes, for instance, or will want to use a cryptographically stronger seed. All of this shoud be done under the application itself, and definitely not be silently overridden by such an import.

This also touches a different general problem that we discussed when conceiving _x-go_: we must not blindly import code from internal application packages and ignore the design of the API. Writing an internal small function in a rush is very different from the designing a long term API for public consumption. So when importing, let's please consider the overall design of the APIs, and whether we want to recommend their use openly as-is.

Reply:
Yes, I definitely agree it's not good to have this in an imported package's `init()`. However, it's interesting to note that the Go team is changing that behaviour so that as of Go 1.20 `math/rand` will auto-seed on startup. Proposal and rationale here: https://github.com/golang/go/issues/54880